### ISSUE TITLE
403 Error for alternatehistory and questionablequesting

### ISSUE THREAD

**Issue by ezlinn** [JimmXinu/FanFicFare#244]
The fanficfare plugin on calibre gives a 403 error for any site that requires a login. I have all login details saved to my personal.ini (even tried putting it under both [defaults] and [epub]) but the same error keeps popping up.

Sites I tested were alternatehistory and questionablequesting. 

> calibre, version 3.7.0
> ERROR: Unhandled exception: <b>HTTPError</b>:HTTP Error 403: Forbidden
> 
> calibre 3.7 [64bit]  embedded-python: True is64bit: True
> Windows-10-10.0.15063-SP0 Windows ('64bit', 'WindowsPE')
> ('Windows', '10', '10.0.15063')
> Python 2.7.12+
> Windows: ('10', '10.0.15063', 'SP0', u'Multiprocessor Free')
> Interface language: None
> Successfully initialized third party plugins: EpubMerge (2, 1, 1) && FanFicFare (2, 18, 0)
> Traceback (most recent call last):
>   File "calibre_plugins.fanficfare_plugin.fff_plugin", line 548, in get_urls_from_page_menu
>   File "calibre_plugins.fanficfare_plugin.fff_plugin", line 568, in get_urls_from_page
>   File "calibre_plugins.fanficfare_plugin.fanficfare.geturls", line 70, in get_urls_from_page
>   File "calibre_plugins.fanficfare_plugin.fanficfare.configurable", line 1221, in _fetchUrl
>   File "calibre_plugins.fanficfare_plugin.fanficfare.adapters.base_xenforoforum_adapter", line 100, in _fetchUrlOpened
>   File "calibre_plugins.fanficfare_plugin.fanficfare.configurable", line 1210, in _fetchUrlOpened
>   File "calibre_plugins.fanficfare_plugin.fanficfare.configurable", line 1111, in _fetchUrlOpened
> HTTPError: HTTP Error 403: Forbidden

**Comment by JimmXinu** [comment]
I'm able to download from alternatehistory and questionablequesting using the plugin just fine.

Please provide some story URLs that demonstrate the issue you're having.

**Comment by ezlinn** [comment]
It's pretty much any story that requires login, as far as I can tell. I tried with a random few after the ones I initially wanted failed.

I just tried this one, for e.g.

https://www.alternatehistory.com/forum/threads/not-throwin-away-my-shot-or-the-adventures-of-alexander-hamilton-in-westeros.422504/

**Comment by JimmXinu** [comment]
I'm able to download that story with `username` and `password` from `personal.ini`.  I'm also prompted and able to enter user/pass if I have a bad password saved.

Have you been downloading tons of stories from those sites recently?  Your IP or user/pass might be specifically throttled or blocked.

We've seen anti-virus/software firewalls running on the client computer block or corrupt FFF's outbound connections before, but the behavior was different.

**Comment by ezlinn** [comment]
I have never downloaded any story from these sites because they were all unsuccessful. I just put calibre on Windows Firewall exceptions and still got the same error.

Is there anything telling in the error I get? 

I really don't understand why it's failing, especially to even get a prompt. I tried to use the Fanficfare app online (fanficfare.appspot.com) and that didn't seem to work either after I input my login to User Config. I feel like I'm missing something basic.

Calibre steps: Configure fanficfare > personal.ini tab > Edit personal.ini > copy of below under both [defaults] and [epub] with no # in front of lines or red text indicating errors

[www.alternatehistory.com]
username:myuser
password:mypw
is_adult:true

**Comment by JimmXinu** [comment]
I don't know either.  If you run calibre in debug mode (Calibre's Preferences dropdown), try again and then post the full debug log it might help tell me what's wrong.

**Comment by ezlinn** [comment]
> calibre Debug log
> calibre 3.10 [64bit]  embedded-python: True is64bit: True
> Windows-10-10.0.15063-SP0 Windows ('64bit', 'WindowsPE')
> ('Windows', '10', '10.0.15063')
> Python 2.7.12+
> Windows: ('10', '10.0.15063', 'SP0', u'Multiprocessor Free')
> Interface language: None
> Successfully initialized third party plugins: FanFicFare (2, 18, 0) && EpubMerge (2, 1, 1)
> calibre 3.10 [64bit]  embedded-python: True is64bit: True
> Windows-10-10.0.15063-SP0 Windows ('64bit', 'WindowsPE')
> ('Windows', '10', '10.0.15063')
> Python 2.7.12+
> Windows: ('10', '10.0.15063', 'SP0', u'Multiprocessor Free')
> Interface language: None
> Successfully initialized third party plugins: FanFicFare (2, 18, 0) && EpubMerge (2, 1, 1)
> Turning on automatic hidpi scaling
> devicePixelRatio: 1.0
> logicalDpi: 96.0 x 96.0
> physicalDpi: 95.8113948919 x 95.9160839161
> Using calibre Qt style: True
> Starting up...
> FFF: DEBUG: 2017-10-29 15:46:54,253: calibre_plugins.fanficfare_plugin.fff_plugin(207): Plugin FanFicFare macmenuhack file_path:C:\Users\MYPC\AppData\Roaming\calibre\plugins\fanficfare_macmenuhack.txt
> EpubMerge: DEBUG: 2017-10-29 15:46:54,325: calibre_plugins.epubmerge.epubmerge_plugin(152): macmenuhack file_path:C:\Users\PC\MYPC\Roaming\calibre\plugins\fanficfare_macmenuhack.txt
> Started up in 13.02 seconds with 101 books
> FFF: DEBUG: 2017-10-29 15:47:25,345: calibre_plugins.fanficfare_plugin.fff_plugin(562): get_urls_from_page URL:https://www.alternatehistory.com/forum/threads/not-throwin-away-my-shot-or-the-adventures-of-alexander-hamilton-in-westeros.422504/
> FFF: DEBUG: 2017-10-29 15:47:25,349: calibre_plugins.fanficfare_plugin.fanficfare.configurable(1038): #####################################
> pagecache(GET) MISS: https://www.alternatehistory.com/forum/threads/not-throwin-away-my-shot-or-the-adventures-of-alexander-hamilton-in-westeros.422504/
> FFF: DEBUG: 2017-10-29 15:47:25,434: calibre_plugins.fanficfare_plugin.fanficfare.configurable(1103): Caught an exception reading URL: https://www.alternatehistory.com/forum/threads/not-throwin-away-my-shot-or-the-adventures-of-alexander-hamilton-in-westeros.422504/  Exception HTTP Error 403: Forbidden.
> FFF: DEBUG: 2017-10-29 15:47:25,434: calibre_plugins.fanficfare_plugin.fanficfare.configurable(1109): Giving up on https://www.alternatehistory.com/forum/threads/not-throwin-away-my-shot-or-the-adventures-of-alexander-hamilton-in-westeros.422504/
> FFF: DEBUG: 2017-10-29 15:47:25,434: calibre_plugins.fanficfare_plugin.fanficfare.configurable(1110): HTTP Error 403: Forbidden
> Traceback (most recent call last):
>   File "calibre_plugins.fanficfare_plugin.fanficfare.configurable", line 1098, in _fetchUrlOpened
>   File "calibre_plugins.fanficfare_plugin.fanficfare.configurable", line 1059, in _fetchUrlRawOpened
>   File "urllib2.py", line 435, in open
>   File "urllib2.py", line 548, in http_response
>   File "urllib2.py", line 473, in error
>   File "urllib2.py", line 407, in _call_chain
>   File "urllib2.py", line 556, in http_error_default
> HTTPError: HTTP Error 403: Forbidden
> calibre 3.10 [64bit]  embedded-python: True is64bit: True
> Windows-10-10.0.15063-SP0 Windows ('64bit', 'WindowsPE')
> ('Windows', '10', '10.0.15063')
> Python 2.7.12+
> Windows: ('10', '10.0.15063', 'SP0', u'Multiprocessor Free')
> Interface language: None
> Successfully initialized third party plugins: FanFicFare (2, 18, 0) && EpubMerge (2, 1, 1)
> Traceback (most recent call last):
>   File "calibre_plugins.fanficfare_plugin.fff_plugin", line 548, in get_urls_from_page_menu
>   File "calibre_plugins.fanficfare_plugin.fff_plugin", line 568, in get_urls_from_page
>   File "calibre_plugins.fanficfare_plugin.fanficfare.geturls", line 70, in get_urls_from_page
>   File "calibre_plugins.fanficfare_plugin.fanficfare.configurable", line 1221, in _fetchUrl
>   File "calibre_plugins.fanficfare_plugin.fanficfare.adapters.base_xenforoforum_adapter", line 100, in _fetchUrlOpened
>   File "calibre_plugins.fanficfare_plugin.fanficfare.configurable", line 1210, in _fetchUrlOpened
>   File "calibre_plugins.fanficfare_plugin.fanficfare.configurable", line 1111, in _fetchUrlOpened
> HTTPError: HTTP Error 403: Forbidden

**Comment by JimmXinu** [comment]
I suppose it's my own fault I didn't notice the initial log showed that you were actually trying to do `get_urls_from_page`(_Get Story URLs from Web Page_) and not a simple story download(_Download from URLs_) as I assumed.  

OTOH, _why_ are you trying to get a list of story URLs from a thread URL?  It doesn't look like there are any story URL list posts there.  If your intention is to download a story using [this index post](https://www.alternatehistory.com/forum/threads/not-throwin-away-my-shot-or-the-adventures-of-alexander-hamilton-in-westeros.422504/#post-15291690), use that URL as the story URL.

There is already special code in the `get_urls_from_page` function for logging into AO3, but I'd like to avoid adding more for other sites.  So I'm going to want to know exactly why you want that before I'll consider changing the code.

**Comment by ezlinn** [comment]
Damn it. Ingrained habits. I developed the habit from a long time ago when I first started downloading threads from SB/SV and in the past I always noticed that downloading via (Download from URLs) gave a horribly formatted and faulty epub. Such as missing the first post (sometimes the index, but often including the first chapter), chapters all being named "First Post" rather than by their title, thereby ruining the TOC, and files being named with a mix of "OEBPS/file0001.xhtml" or "3/OEBPS/file0001.xhtml" rather than an orderly "file0001.xhtml, file0002.xhtml, etc". 

I used an improvised solution: use the Threadmarks link, (Get Story URLs), then manually copy/paste out the relevant threadmarked links and add the first threadmark (usually the thread link) on top, before pasting it all back and downloading. This results in the faulty epub files duplicated after the clean files. I delete those by editing and get the clean version. 

I just used (Download from URLs) and there's no issue with the output now. AH and QQ links both worked.

Thanks for the troubleshooting!
